### PR TITLE
[Caching] Fix the cache replay bugs from driver C APIs

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
This patch contains a serials of changes that targeted to fix the missing function in C APIs when replaying results from build system. It teaches the driver/build-system to replay the newly specialized makefile dependency files and symbol graph files correctly.

It also tags along and fixes few bugs and small improvements that intertwined with the bug fix that hardly be separated so they comes in separated commits in this PR.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Review by rastogishubham — APPROVED** [review]
I am not familiar with the symbol-graph code, but the MCCAS code LGTM!

**Comment by cachemeifyoucan** [comment]
@benlangmuir for symbol graph changes because I missed that from the code review.

File: lib/Tooling/libSwiftScan/SwiftCaching.cpp [hunk 1]
```diff
@@ -956,15 +961,38 @@ static llvm::Error replayCompilation(SwiftScanReplayInstance &Instance,
             if (!Proxy)
               return Proxy.takeError();
 
-            if (Kind == file_types::ID::TY_Object && UseCASBackend)
-              ObjFile = OutputPath->second;
-
             if (Kind == file_types::ID::TY_CachedDiagnostics) {
               assert(!DiagnosticsOutput && "more than 1 diagnostics found");
               DiagnosticsOutput = std::move(*Proxy);
+            } else if (Kind == file_types::ID::TY_SymbolGraphFile &&
```
Comment (benlangmuir) [inline lib/Tooling/libSwiftScan/SwiftCaching.cpp:967]: Why is this replay code duplicated?

Comment (cachemeifyoucan) [reply lib/Tooling/libSwiftScan/SwiftCaching.cpp:967]: The historical reason is the diagnostics handling part of the replay is really different so this is a separate copy.

I am working on the follow up to eliminate the copy as an improvement. Some handling will be really hard to unify but I think moving the main replay logics into the same file and keep only one copy is the way to go.

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**